src/hexadecimal.jl

```diff
@@ -0,0 +1,160 @@
+            h = ((h << 4) | c)
+            pos += 1
+            incr!(source)
+        end
+        # Only check validity every 8 bytes to hopefully enable some vectorization
+        if check == 0xFFFF
+            pos -= 8 # Backtrack to the start of the invalid hex
+            fastseek!(source, pos)
+            while _HEX_LUT[peekbyte(source, pos) + 0x01] != 0xFFFF
+                pos += 1
+                incr!(source)
+            end
+            return pos, code | INVALID, PosLen(pl.pos, pos - pl.pos), SHA1(out)
+        end
+        out = _set_tup5_at(out, i, h)
+    end
+    eof(source, pos, len) && (code |= EOF)
+    return (pos, code | OK, PosLen(pl.pos, pos - pl.pos), SHA1(out))
+end
+
+### UUID ###
+
+supportedtype(::Type{UUID}) = true
+
+function typeparser(::Type{UUID}, source, pos, len, b, code, pl, options)
+    check = UInt32(0)
+    hi = UInt64(0)
+    lo = UInt64(0)
+
+    @inbounds begin
+    if len - pos + 1 < 36
+        while pos <= len && (_HEX_LUT[b + 0x01] != 0xFFFF || b == UInt8('-'))
+            pos += 1
+            incr!(source)
+            b = peekbyte(source, pos)
+        end
+        eof(source, pos, len) && (code |= EOF)
+        return (pos, code | INVALID, PosLen(pl.pos, pos - pl.pos), UUID((hi, lo)))
+    end
+
+    segment_len = 8
+    for _ in 1:segment_len
+        b = peekbyte(source, pos)
+        c = _HEX_LUT[b + 0x01]
+        check |= c
+        hi = ((hi << 4) | UInt32(c))
+        pos += 1
+        incr!(source)
+    end
+    check != 0xFFFF || @goto backtrack_error
+    peekbyte(source, pos) == UInt8('-') || @goto error
+
+    pos += 1
+    incr!(source)
+    segment_len = 4
+    for _ in 1:segment_len
+        b = peekbyte(source, pos)
+        c = _HEX_LUT[b + 0x01]
+        check |= c
+        hi = ((hi << 4) | UInt32(c))
+        pos += 1
+        incr!(source)
+    end
+    check != 0xFFFF || @goto backtrack_error
+    peekbyte(source, pos) == UInt8('-') || @goto error
+
+    pos += 1
+    incr!(source)
+    segment_len = 4
+    for _ in 1:segment_len
+        b = peekbyte(source, pos)
+        c = _HEX_LUT[b + 0x01]
+        check |= c
+        hi = ((hi << 4) | UInt32(c))
+        pos += 1
+        incr!(source)
+    end
+    check != 0xFFFF || @goto backtrack_error
+    peekbyte(source, pos) == UInt8('-') || @goto error
+
+    pos += 1
+    incr!(source)
+    segment_len = 4
+    for _ in 1:segment_len
+        b = peekbyte(source, pos)
+        c = _HEX_LUT[b + 0x01]
+        check |= c
+        lo = ((lo << 4) | UInt32(c))
+        pos += 1
+        incr!(source)
+    end
+    check != 0xFFFF || @goto backtrack_error
+    peekbyte(source, pos) == UInt8('-') || @goto error
+
+    pos += 1
+    incr!(source)
+    segment_len = 12
+    for _ in 1:segment_len
+        b = peekbyte(source, pos)
+        c = _HEX_LUT[b + 0x01]
+        check |= c
+        lo = ((lo << 4) | UInt32(c))
+        pos += 1
+        incr!(source)
+    end
+    check != 0xFFFF || @goto backtrack_error
```

Comment:
can these be functions rather than `@goto`s?

e.g. 
```suggestion
    check != 0xFFFF ||  return backtrack_error!(source, pos, segment_len, ...)
```

Reply:
So I tried a couple things here:
a) current thing where jump to the error handling segment of the code
b) make a `@inline`d function `_backtrack_error(source, pos, len, code, segment_len, hi, lo, pl)`
c) make a `@noinline`d function `_backtrack_error(source, pos, len, code, segment_len, hi, lo, pl)`

In an micro benchmark where we parse a single valid UUID string, I didn't see any perf difference. But there is a difference in the size of the generated code. For the three versions I ran

```julia
@code_native debuginfo=:none syntax=:intel Parsers.typeparser(UUID, collect(codeunits(su)), 1, (length(su)), 0x00, Int16(0), Parsers.PosLen(0,0,false,false), (Parsers.OPTIONS))
```
And then I very scientifically counted the number of lines generated:
a) 488
b) 705
c) 507

So I think there the original version is at least preferable in terms of codesize? The `@noinline`d seems to only differ in that it inserts a couple of  `call` instructions which are also just gotos/jumps + some small changes to the stack (if we hit the error, that is). But do let me know if you think that the gains in legibility are worth it.

Reply:
i personally this the gains in legibility are worth it if there's no time cost